src/physfs_tar.h

```diff
@@ -0,0 +1,299 @@
+	TSVTX            = 1 << 9, /* reserved */
+	TSGID            = 1 << 10, /* set GID on execution */
+	TSUID            = 1 << 11, /* set UID on execution */
+};
+
+/* POSIX header.  */
+struct posix_header
+{				  /* byte offset */
+	char name[100];           /*   0 */
+	char mode[8];             /* 100 */
+	char uid[8];              /* 108 */
+	char gid[8];              /* 116 */
+	char size[12];            /* 124 */
+	char mtime[12];           /* 136 */
+	char chksum[8];           /* 148 */
+	char typeflag;            /* 156 */
+	char linkname[100];       /* 157 */
+	char magic[TMAGLEN];      /* 257 */
+	char version[2];          /* 263 */
+	char uname[32];           /* 265 */
+	char gname[32];           /* 297 */
+	char devmajor[8];         /* 329 */
+	char devminor[8];         /* 337 */
+	char prefix[155];         /* 345 */
+	                          /* 500 */
+};
+
+/* tar Header Block, from POSIX 1003.1-1990.  */
+union block
+{
+  char buffer[BLOCKSIZE];
+  struct posix_header header;
+};
+
+static PHYSFS_uint64 TAR_decodeOctal(char* data, size_t size) {
+    unsigned char* currentPtr = (unsigned char*) data + size;
+    PHYSFS_uint64 sum = 0;
+    PHYSFS_uint64 currentMultiplier = 1;
+    unsigned char* checkPtr = currentPtr;
+
+    for (; checkPtr >= (unsigned char*) data; checkPtr--) {
+        if ((*checkPtr) == 0 || (*checkPtr) == ' ') {
+            currentPtr = checkPtr - 1;
+        }
+    }
+    for (; currentPtr >= (unsigned char*) data; currentPtr--) {
+        sum += (*currentPtr - 48) * currentMultiplier;
+        currentMultiplier *= 8;
+    }
+    return sum;
+}
+
+static bool TAR_encodeOctal(char* data, size_t size, PHYSFS_uint64 i) {
+    if(	snprintf(data, size, "%llo", i) < 0 )
+    	return false;
+    return true;
+}
+
+static int TAR_magic(union block* block) {
+	if(strcmp(block->header.magic, OLDGNU_MAGIC) == 0 && strncmp(block->header.version, TVERSION, 2) == 0)
+		return (int)OLDGNU_FORMAT;
+	if(strncmp(block->header.magic, TMAGIC, TMAGLEN - 1) == 0)
+		return (int)POSIX_FORMAT;
+	return (int)DEFAULT_FORMAT;
+}
+
+static size_t TAR_fileSize(union block* block) {
+	return TAR_decodeOctal(block->header.size, sizeof(block->header.size));
+}
+
+static bool TAR_checksum(union block* block) {
+	PHYSFS_sint64 unsigned_sum = 0;
+	PHYSFS_sint64 signed_sum = 0;
+	PHYSFS_uint64 reference_chksum = 0;
+	char orig_chksum[8];
+	int i = 0;
+
+	memcpy(orig_chksum, block->header.chksum, 8);
+	memset(block->header.chksum, ' ', 8);
+
+	for(; i < BLOCKSIZE; i++) {
+		unsigned_sum += ((unsigned char*) block->buffer)[i];
+			signed_sum += ((signed char*) block->buffer)[i];
+	}
+	memcpy(block->header.chksum, orig_chksum, 8);
+	reference_chksum = TAR_decodeOctal(orig_chksum, 12);
+	return (reference_chksum == unsigned_sum || reference_chksum == signed_sum);
+}
+
+time_t TAR_time(union block* block)
+{
+	return TAR_decodeOctal(block->header.mtime, 12);
+}
+
+mode_t TAR_mode(union block* block)
```

Comment:
The mode functions `TAR_mode` and `TAR_set_mode` aren't used at all: can be commented out. (also see above.)